### ISSUE TITLE
training with missing keypoints?

### ISSUE THREAD

**Issue by Shreyas-NR** [zhangboshen/A2J#61]
Hi @zhangboshen,
I'm planning to retrain the A2J model with my custom dataset to predict Human Pose. 
I'm annotating the dataset and I stumbled into a situation where a frame does not contain a part of the human body(eg. head, neck, and Left Shoulder). If I do not annotate a joint, then its pixel coordinate would be (x=0,y=0). 

Could you please advise me on what should I do in this case for the model to predict the rest of the joints accurately and not get influenced by missing keypoint joints?

**Comment by zhangboshen** [comment]
In such cases, I suggest add a confidence output to indicate if a joint is visible. And the annotations may ignore the invisible joints. In training, these invisiable joints can be discarded and should not contribute any loss.

**Comment by Shreyas-NR** [comment]
Hi @zhangboshen,
Thanks for your reply,
Excellent idea of adding a confidence output to indicate the visibility of the joint. I will work on it in my annotation exporting. 
I am planning to use this logic for visibility flag v defined as, 

> v=0: not labeled (in which case x=y=0), 
> v=1: labeled but not visible, and 
> v=2: labeled and visible
```
    keypoints = [
        'head',
        'neck',
        'right_shoulder',
        'left_shoulder',
        'right_elbow',
        'left_elbow',
        'right_wrist',
        'left_wrist',
        'torso',
        'right_hip',
        'left_hip',
        'right_knee',
        'left_knee',
        'right_ankle',
        'left_ankle']
```
![image](https://user-images.githubusercontent.com/58805235/227386541-1d0aa9f6-7ab3-4e89-a336-2140dc7b6a59.png)


However, I'm not very clear when you say "In training, these invisible joints can be discarded and should not contribute any loss."

Before the training loop, we execute the torch.dataloadaer that reads the image(crop it to 288x288) and its label(15 keypoints x 3)
https://github.com/zhangboshen/A2J/blob/4c332f0ebff159d384af162a90b64ce89ea8d4b4/src_train/nyu.py#L183

> `    label_xy[:, 0] = (keypointsPixel[index, :, 0] - new_Xmin) * cropWidth / (new_Xmax - new_Xmin)  # x   
>     label_xy[:, 1] = (keypointsPixel[index, :, 1] - new_Ymin) * cropHeight / (new_Ymax - new_Ymin)  # y`

then the image is processed through the A2J network and loss is computed using

```
> criterion = anchor.A2J_loss(shape=[cropHeight//16,cropWidth//16],thres = [16.0,32.0],stride=16,\
>         spatialFactor=spatialFactor,img_shape=[cropHeight, cropWidth],P_h=None, P_w=None)
> heads  = net(img) 
> Cls_loss, Reg_loss = criterion(heads, label)
```
https://github.com/zhangboshen/A2J/blob/4c332f0ebff159d384af162a90b64ce89ea8d4b4/src_train/anchor.py#L84


Can you please advise me on where should I add the condition to ignore the invisible joints? I understand it's been a long time since you have written this code, any pointers will be very helpful. 


thanks in advance,

**Comment by zhangboshen** [comment]
@Shreyas-NR 
I think the invisible joints should not contribute any regression loss during training and it can be done by multiply the per-joint loss with a 0/1 mask, where 0 indicates invisible (or not annotated) and 1 indicates the visible joint.
And as for the confidence branch, the 0/1 can also work as the GT label for supervision.
